docs(snapcraft): aggregate lifecycle docs into an explanation

Sophie-Pages opened the pull request:
First step into moving the Parts lifecycle into Snapcraft's reference: create the reference.
It is missing the additional work and redirect links on the other Snapcraft pages.

dilyn-corner commented on snapcraft/explanation/parts-lifecycle.md:
```diff
@@ -0,0 +1,90 @@
+Each plugin defines the default actions that happen during a step. This behavior can be changed in two ways.
+
+- By using `override-<step-name>` in `snapcraft.yaml`. See [Overriding steps](/t/scriptlets/4892) for more details.
+- By using a local plugin.  This can inherit the parent plugin or scaffolding from the original. See [Local plugins](/t/writing-local-plugins/5125) for more details.
```
It may be prudent to mention that this hasn't been supported for recent bases in snaps for several years -- I believe the cut-off is core20?

dilyn-corner commented on snapcraft/explanation/parts-lifecycle.md:
```diff
@@ -0,0 +1,90 @@
+| `CRAFT_PART_SRC` | **`parts/<part-name>/src`** | the location of the source during the *pull* step |
+| `CRAFT_PART_BUILD` | **`parts/<part-name>/build`** | the working directory during the *build* step |
+| `CRAFT_PART_INSTALL`| **`parts/<part-name>/install`** | contains the results of the *build* step and the stage packages. |
+| `CRAFT_STAGE` | **`stage`** | shared by all parts, this directory contains the development libraries, headers, and other components (e.g.; pkgconfig files) that need to be accessible from other parts |
```
"This directory *can* contain" -- it isn't necessary that they do so (depends on the plugin, `stage:`, the build script in the source itself, actions in `override-build`, etc.)
Likewise it *definitely* contains the contents of each part's `CRAFT_PART_INSTALL` sans any likewise shenanigans as above.

dilyn-corner commented on snapcraft/how-to/quickstart/how-snapcraft-builds-a-snap.md:
```diff
@@ -0,0 +1,90 @@
+
+These operations translate into practical tasks like:
+
+* It can build snaps locally or send remote tasks to Launchpad.
+* It allows the developer to register and log into the Snap Store.
+* It can upload snaps to the Snap Store.
+* It can promote snaps to different channels.
```
I would remove the "It" from these lines; it's well established we're talking about Snapcraft, and it better mirrors the previous lines (4-6).

dilyn-corner commented on snapcraft/how-to/quickstart/how-snapcraft-builds-a-snap.md:
```diff
@@ -0,0 +1,90 @@
+
+Snaps are created using a build recipe defined in a file called `snapcraft.yaml`.
+
+When the snapcraft tool is executed on the command line, it will look for the file in the current project work directory, either in the top-level folder or a `snap` subdirectory. If the file is found, snapcraft will then parse its contents and progress the build toward completion.
```
@medubelko does snapcraft support `build-aux/snap`? Is it a recommended path?

medubelko replied:
I'm not familiar with this feature, and can't find much info about it. @mr-cal would know for certain.

With the remaining migration to RTD scheduled for the end of cycle, it's likely that much of this page will be deemed out of date and retired.

dilyn-corner commented on snapcraft/how-to/quickstart/how-snapcraft-builds-a-snap.md:
```diff
@@ -0,0 +1,90 @@
+
+<h3 id='heading--definitions'>Main definitions inside snapcraft.yaml</h3>
+
+There is no one way for how a snap ought to be assembled. However, most `snapcraft.yaml` files have the same common elements, including a number of mandatory declarations. Below is a short list of these keys, which will be further explained in the Examples sections later in the tutorial.
```
As we call out that only *some* keys are mandatory, should we specify in 46-50 which of them are mandatory? Parts and Apps are both optional, for instance - but one wouldn't know it from reading this.

dilyn-corner commented on snapcraft/how-to/quickstart/how-snapcraft-builds-a-snap.md:
```diff
@@ -0,0 +1,90 @@
+
+It is important to note several additional details:
+
+* A snap may contain one or more parts.
+* A snap may contain one or more applications
```
"None or more", especially for apps -- snaps like `mesa-core22` (and earlier/later iterations) don't include any apps, but merely *serve content* via a content interface.

dilyn-corner commented on snapcraft/how-to/quickstart/how-snapcraft-builds-a-snap.md:
```diff
@@ -0,0 +1,90 @@
+* A snap may contain one or more applications
+* Parts can be pre-assembled binaries or they may be compiled as part of the build process.
+* The parts section of the `snapcraft.yaml` file uses Snapcraft build system or language-specific plugins to simplify the build process.
+* The parts section may also include a list of [build packages](/t/build-and-staging-dependencies/11451) (build-packages) that will be used to create the snap applications but will not be included in the final snap. For instance, gcc or make.
+
+The parts section may also include a list of [stage packages](/t/build-and-staging-dependencies/11451) (stage-packages) that will be used by the snap’s applications at runtime, e.g.: python-bcrypt. These will be obtained from the repository archives in the build instance.
```
Should we give a link to adding package repositories after these lines?

dilyn-corner commented on snapcraft/explanation/parts-lifecycle.md:
```diff
@@ -0,0 +1,90 @@
+The steps of a part goes through are as follows:
+
+1. **pull**: downloads or otherwise retrieves the components needed to build the part. You can use the [`source-*` keywords](/t/snapcraft-parts-metadata/8336#heading--source) of a part to specify which components to retrieve. If `source` points to a git repository, for example, the pull step will clone that repository.
+1. **build**: constructs the part from the previously pulled components. The [`plugin`](/t/snapcraft-plugins/4284) of a part specifies how it is constructed. The [`meson` plugin](/t/the-meson-plugin/8623), for example, executes `meson` and `ninja` to compile source code. Each part is built in a separate directory, but it can use the contents of the staging area if it specifies a dependency on oth ...
```
I will note that `organize` is missing -- considering its importance and relationship to the `stage` and `prime` keywords, it should be included here.

medubelko replied:
We have a problem in the existing docs where they imply that `organize` it itself a step, rather than a before/after events. `organize` belongs to Stage, not Build, so I think we should be careful about where we lodge the keys and how we organise the lifecycle.

One way we could model and present the timeline is with – to use an idea from music – _on beat_ events for steps (**one** and **two** and) and _off beat_ events for the in-between procedures (one **and** two **and**). Shown as a tree:

1. Pull
    1. `after`
    2. etc
2. Overlay
    1. `overlay-script` (to borrow a case from Rockcraft)
    2. etc
4. Build
    1. `build-packages`
    2. `build-snaps`
    3. etc

For absolute correctness, it would probably need to be presented as a table that resembles a gantt-like schedule.

If we did this, whether we include _all_ the keys remains a question. There are many. I think Sophie is already on the right track by including the most important in-between events and providing examples.

Do you think this could be a helpful way to map the steps?

dilyn-corner replied:
I'm not familiar with such a style, but it's potentially a helpful way of organizing this information -- at the very least, I wouldn't see any harm in attempting adding a structure like this!

Is it not the case that `organize` belongs to the build step? I believe `organize`d bits will end up modified in the part's `$CRAFT_PART_INSTALL`, which then *carries over* into `$CRAFT_STAGE`.

medubelko replied:
> Is it not the case that `organize` belongs to the build step?

I might be misled. `organize` is described as both part of the [stage step](https://canonical-craft-parts.readthedocs-hosted.com/en/latest/common/craft-parts/reference/part_properties.html#organize) and the [build step](https://canonical-craft-parts.readthedocs-hosted.com/en/latest/common/craft-parts/reference/part_properties.html#summary-of-keys-and-steps).

@mr-cal Which is correct?

mr-cal replied:
`organize` is the last event in the build step.  It moves files within `$CRAFT_PART_INSTALL`.  

> I believe organized bits will end up modified in the part's $CRAFT_PART_INSTALL, which then carries over into $CRAFT_STAGE. 

@dilyn-corner is correct here.

dilyn-corner reviewed (commented):
I really like this -- good work! Left some comments, mostly high-level thoughts (those ending with '?'). Others are, I think, more necessary changes.

Sophie-Pages commented:
Thank you @dilyn-corner and @medubelko for your precious comments! I will take all that into account and create new commits very soon!

Sophie-Pages commented:
Hi @dilyn-corner and @medubelko,

I reviewed your comments except for the support build-aux/snap comment.

After talking with Michael we decided that I could try to improve the lifecycle explanation even if it's out of the scope of the initial issue.
I cleaned, clarified, reorganised, and added missing information related to the lifecycle. 

Two topics remain on my side:
- I added information about permissions in the prime step. I think the users would benefit from having a link to more info on this topic but I don't think the lifecycle explanation is the place to do that. 
Can I link the [readthedocs ](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/reference/parts_steps/) page and maybe in another issue a snapcraft.io page on permissions could be created?
- I would like to add a [similar processing diagram](https://canonical-craft-parts.readthedocs-hosted.com/en/latest/common/craft-parts/explanation/lifecycle.html#lifecycle-processing-diagram). Because it will need to be redone from sratch, I wanted your approval before working on it.

Is there anything else that you think could be missing from the lifecycle explanation?

Thank you

dilyn-corner commented on snapcraft/explanation/parts-lifecycle.md:
```diff
@@ -0,0 +1,147 @@
+    * This is the first time all the different parts that make up the snap are actually placed in the same directory. If multiple parts provide the same file with differing contents, you will get a conflict. You can avoid these conflicts by using the [`stage` keyword](/t/snapcraft-parts-metadata/8336#heading--stage) to enable or block files coming from the part. You can also use this keyword to f ...
+1. **Prime**: copies the staged components into the priming area. 
+    * This is very similar to the stage step, but files go into the priming area instead of the staging area. The `prime` step exists because the staging area might still contain files that are required for the build but not for the snap. For example, if you have a part that downloads and installs a compiler, then you stage this part so other parts can use the compiler during building. You can th ...
+    * During this step, if specific `permissions` were specified, they will be applied.
```
What exactly is meant by 'permissions' here? I'm not familiar with this aspect of the prime step.

Sophie-Pages replied:
I suppose it is indeed more of a "craft" thing but because there is an existing [documentation](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/reference/parts_steps/#permissions) I thought it would be useful to mention it for the users. 
It works just fine for snaps (I tested it) and allows to change the permissions of specific files during the prime step.

dilyn-corner commented on snapcraft/explanation/parts-lifecycle.md:
```diff
@@ -0,0 +1,147 @@
+
+Each plugin defines the default actions that happen during a step. This behavior can be changed in two ways:
+- By using `override-<step-name>` in `snapcraft.yaml`. See [Overriding steps](/t/scriptlets/4892) for more details.
+- By using a local plugin.  This can inherit the parent plugin or scaffolding from the original. See [Local plugins](/t/writing-local-plugins/5125) for more details. Please note that using a local plugin is deprecated for base core22 and onwards. Because it hasn't been supported for recent bases, it will need a core20 to work.  
```
> Please note that using a local plugin is deprecated for base core22 and onwards. Because it hasn't been supported for recent bases, it will need a core20 to work.  

This phrasing feels a little cludgy; local plugins aren't just deprecated, they're wholly unsupported (albeit, there's technically some tricks you can pull, but we don't mention that...). "it will need a core20" should probably read "the snap must use `base: core20` or earlier".

Sophie-Pages replied:
```suggestion
- By using a local plugin.  This can inherit the parent plugin or scaffolding from the original. See [Local plugins](/t/writing-local-plugins/5125) for more details. Please note that using a local plugin has been deprecated in `base: core20` and fully removed in `base: core22`.  
```
I chose to frame it this way so the users are not led to think they should use core20 while 20.04 will reach end of free support in a few months. What do you think?

Sophie-Pages replied:
For the `organize` link, do you want me to pull the [reference ](https://snapcraft.io/docs/snapcraft-yaml-schema#organize-79) in CODA and add an anchor? Maybe each key from this page would benefit of having an anchor but that would be a different issue I suppose? I can as well use this link https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-organize-79, even if it is not ideal, that would be a middle ground before the issue on anchors is done? Either way, I can do what you think is best.

dilyn-corner replied:
The new language is great!

I agree that each key should have its own anchor (why do to the sidebar hyperlinks not include links to the specific parts directives, for instance?), and I agree that doing this would be more apt as a separate issue. But I think at that point we are fast approaching a design consideration that I don't think I'm qualified for; @degville @medubelko?

medubelko replied:
@Sophie-Pages 

> For the organize link, do you want me to pull the [reference ](https://snapcraft.io/docs/snapcraft-yaml-schema#organize-79)in CODA and add an anchor? Maybe each key from this page would benefit of having an anchor but that would be a different issue I suppose?

There's no need to add more pages to this branch; I can sort out the particulars of links between pages when I move this to Discourse. :)

dilyn-corner reviewed (approved):
> I added information about permissions in the prime step. I think the users would benefit from having a link to more info on this topic but I don't think the lifecycle explanation is the place to do that.

Agreed; I also haven't ever really seen documentation surrounding this, which gives me the impression that it's more of a general "craft" thing than a snap thing (is it for rocks? charms?).

> Can I link the [readthedocs ](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/reference/parts_steps/)page and maybe in another issue a snapcraft.io page on permissions could be created?

Ultimately I think it may fall to @mr-cal's judgement; I've never seen a snap use this syntax.

> I would like to add a [similar processing diagram](https://canonical-craft-parts.readthedocs-hosted.com/en/latest/common/craft-parts/explanation/lifecycle.html#lifecycle-processing-diagram). Because it will need to be redone from sratch, I wanted your approval before working on it.

I think this would be a very powerful explanatory tool, but I also don't think it should block merging these excellent changes.

> Is there anything else that you think could be missing from the lifecycle explanation?

It may be useful to add a link to `organize`'s definition or use similarly to how it's done for the `stage` keyword; I think at the very least a link to [here](https://snapcraft.io/docs/snapcraft-yaml-schema#organize-79) (and that page having anchors would be super useful...).

Sophie-Pages commented:
Hi @dilyn-corner and @medubelko ,

- I created and added the processing diagram, please tell me if it is correct or if I need to make some adjustments.
- I added this link https://snapcraft.io/docs/snapcraft-yaml-schema#p-21225-organize-79 in the organize event for now.
- I kept the information about permissions but without any link as I am not sure if we want to redirect to the [readthedocs page](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/reference/parts_steps/).

Sophie-Pages commented:
Hi @medubelko,
I added the diagram source file for maintainability.

I believe the diagram could benefit from a written text to enhance accessibility. This could be done in another issue to ensure it doesn't block the PR.

dilyn-corner commented:
This looks fantastic, @Sophie-Pages!

I think there needs to be an "N" on the right arrow flowing from the "Step already ran?" node.

Beyond that, I think this is a fantastic set of changes to merge with some light TODOs for future expansion. @medubelko, thoughts?

medubelko commented on snapcraft/reference/parts-lifecycle.md:
As you may recall from our offline discussion, I now prefer this task to be focused on the explanation. I don't think we need this as a page anymore.

I might copy some of the wording when I port this to Discourse.

medubelko reviewed (approved):
@Sophie-Pages Thanks for putting this together!

The comments that start with "For later" are notes for porting and migration. It will be a two step process where we move this to Discourse first, and then later perform even more of a consolidation before migration to RTD later in the spring.

There's one small comment about Snapcraft 6 that you could look into, and I've pinged the developer to give the steps in the explanation a once-over.

I'm fine with using this work as-is. I'll make tweaks to the language for style, but those don't need to block this PR.

Thanks again!

medubelko commented:
@dilyn-corner 

> I think there needs to be an "N" on the right arrow flowing from the "Step already ran?" node.

This chart is from Craft Parts, and originally from a spec. It has some weaknesses, but it's fine to leave it in as-is for now.

> Beyond that, I think this is a fantastic set of changes to merge with some light TODOs for future expansion. @medubelko, thoughts?

Agreed! I think it's nearly ready for prime time, pending a look from Cal.

@Sophie-Pages 

> I kept the information about permissions but without any link as I am not sure if we want to redirect to the [readthedocs page](https://canonical-snapcraft.readthedocs-hosted.com/en/stable/reference/parts_steps/).

It's fine as-is – no need to sweat about the smaller details, since after I've ported this to Discourse I'll be migrating it to RTD.

Sophie-Pages commented:
Hi @medubelko, 

Thank you for your comments, they were really helpful. I took this chance to improve the explanation and to reduce the smaller details you will need to go through.

I didn't remove the reference, I'll let you decide what you want to do with it.

mr-cal commented on snapcraft/explanation/parts-lifecycle.md:
```diff
@@ -0,0 +1,155 @@
+---
+<h3 id='heading--lifecycle'>Lifecycle<sup><a href=#heading--lifecycle>⚓</a></sup></h3>
+
+A part goes through the following five steps:
+
+1. **Pull**: downloads or otherwise retrieves the components needed to build the part.
+    * You can use the [`source-*`](/t/snapcraft-parts-metadata/8336#heading--source) keywords of a part to specify which components to retrieve. For instance, if `source` points to a Git repository, the pull step will clone that repository.
+1. **Build**: constructs the part from the previously pulled components.
+    * The [`plugin`](/t/snapcraft-plugins/4284) of a part specifies how it is constructed. The [`meson`](/t/the-meson-plugin/8623) plugin, for example, executes `meson` and `ninja` to compile source code.
+    * Each part is built in a separate directory, but it can use the contents of the staging area if it specifies a dependency on other parts using the `after` keyword. See [Processing order and dependencies](#heading--processing-order) for more information.
+    * To rename or move files, you can use the [`organize`](/t/snapcraft-yaml-schema/4276#p-21225-organize-79) keyword.
+1. **Stage**: copies the built components into the staging area.
+    * This is the first time all the different parts that make up the snap are actually placed in the same directory. If multiple parts provide the same file with differing contents, you will get a conflict. You can avoid these conflicts by using the [`stage`](/t/snapcraft-parts-metadata/8336#heading--stage) keyword to enable or block files coming from the part. You can also use this keyword to f ...
+1. **Prime**: copies the staged components into the priming area.
+    * This is very similar to the stage step, but files go into the priming area instead of the staging area. The `prime` step exists because the staging area might still contain files that are required for the build but not for the snap. For example, if you have a part that downloads and installs a compiler, then you stage this part so other parts can use the compiler during building. You can th ...
+    * To apply specific permissions to files, you can use the `permissions` keyword.
+    * Some extra checks are also run during this step to ensure that all dependencies are satisfied for a proper run time. If confinement was set to `classic`, then files will be scanned and, if needed, patched to work with this confinement mode.
+1. **Pack**: takes the entire contents of the `prime` directory and packs it into [a snap](/t/the-snap-format/698).
```
Yep, this looks comprehensive and correct to me!

medubelko replied:
Thanks!

mr-cal commented on snapcraft/explanation/parts-lifecycle.md:
```diff
@@ -0,0 +1,155 @@
+| `CRAFT_PART_INSTALL`| **`parts/<part-name>/install`** | contains the results of the *build* step and the stage packages. It is also the directory where the `organize` event renames the built files |
+| `CRAFT_STAGE` | **`stage`** | shared by all parts, this directory contains the contents of each part's `CRAFT_PART_INSTALL` after the *stage* step. It can contain development libraries, headers, and other components (e.g. pkgconfig files) that need to be accessible from other parts |
+| `CRAFT_PRIME` | **`prime`** | shared by all parts, this directory holds the final components after the *prime* step |
+| `CRAFT_PROJECT_DIR` | path to the current project's subtree in the filesystem. | path to the resulting snap after the *pack* step |
```
I'm not sure the "purpose" section is useful information in this context. `$CRAFT_PROJECT_DIR` only exists when snapcraft is running.  But when snapcraft is running, it doesn't matter where the resulting snap will go.

This variable is more used for accessing a resource in your project directory when the lifecycle is running. For example, [`icon: $CRAFT_PROJECT_DIR/resources/icons/128-apps-ghostwriter.png`](https://github.com/KDE/ghostwriter/blob/6cc5988c4bdd713c1126cfa3235427eb6709b660/snapcraft.yaml#L10).

I can't think of a succinct way to write this, maybe something like:
```suggestion
| `CRAFT_PROJECT_DIR` | path to the current project's subtree in the filesystem. | This is used to access resources from the project's subtree, such as an icon or version file.)|
```

medubelko commented on snapcraft/reference/parts-lifecycle.md:
```diff
@@ -0,0 +1,13 @@
+| **Build** | `snapcraft build [<part-name>]` | Builds the components from the previously pulled sources. | Build the sources in CRAFT_PART_**BUILD** and places the result in CRAFT_PART_**INSTALL**. | **`parts/<part-name>/build`** and **`parts/<part-name>/install`** |
+| **Stage** | `snapcraft stage [<part-name>]` | Copies the built components into the staging area. | The built components are put in CRAFT_**STAGE**. | **`stage`** |
+| **Prime** | `snapcraft prime [<part-name>]` | Copies the staged components into the priming area. | The staged components are put in CRAFT_**PRIME**. | **`prime`** |
+| **Pack** | `snapcraft pack or snapcraft` | Takes the contents of the prime directory and packs it into a snap. | The snap is put in CRAFT_PROJECT_**DIR**. | The path to the current project’s subtree in the filesystem. |
```
We're going in a different direction for this page, so we don't need to make any changes to it.

medubelko commented:
@mr-cal I can't find any info on `snap-aux`. Can you create a ticket for a reference for it and link to any relevant info/code?

medubelko commented:
@Sophie-Pages If you could incorporate Cal's small suggestions, I'll be happy to mark this done and begin the work of bringing it home. :)
